lib/spack/spack/package.py

```diff
@@ -1510,24 +1525,23 @@ def do_install(self, **kwargs):
             # module from the upstream Spack instance.
             return
 
-        partial = self.check_for_unfinished_installation(keep_prefix, restage)
-
         # Ensure package is not already installed
-        layout = spack.store.layout
-        with spack.store.db.prefix_read_lock(self.spec):
-            if partial:
-                tty.msg(
-                    "Continuing from partial install of %s" % self.name)
-            elif layout.check_installed(self.spec):
-                msg = '{0.name} is already installed in {0.prefix}'
-                tty.msg(msg.format(self))
-                rec = spack.store.db.get_record(self.spec)
-                # In case the stage directory has already been created,
-                # this ensures it's removed after we checked that the spec
-                # is installed
-                if keep_stage is False:
-                    self.stage.destroy()
-                return self._update_explicit_entry_in_db(rec, explicit)
+        partial = self.check_for_unfinished_installation(keep_prefix, restage)
```

Comment:
I'm not clear if this rearrangement is relevant to no-source packages specifically, but it does look correct: before there was an issue where two processes could attempt to upgrade from read to write locks and cause a deadlock. Was avoiding that your intent here?

Reply:
Actually that was not my intent.  It just seemed inappropriate to take a lock that was not needed.

Reply:
Do you want me "revert" this change since it isn't specific to no-source packages?

Reply:
Either that (and create a TODO) or move the changes into a separate commit. I think this change is good but it might make it difficult for someone doing a bisect in the future.

Reply:
Okay.

Reply:
FYI. I created a separate PR for this #12478